File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        if(nums == null || nums.length == 0){
            return new ArrayList<>();
        }

        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);

        for(int i = 0; i < nums.length; i++){
            if(i != 0 && nums[i-1] == nums[i]) continue;

            for(int j = i+1; j< nums.length; j++){
                if(j > i+1 && nums[j-1] == nums[j]) continue;

                int k = j+1;
                int l = nums.length-1;

                while(k < l){
                    long sum = nums[i]+nums[j];
                    sum += nums[k];
                    sum += nums[l];

                    if(sum == target){
                        result.add(Arrays.asList(nums[i],nums[j],nums[k],nums[l]));
                        k++;
                        l--;

                        while(k < l && nums[k] ==  nums[k-1]) k++;
                        while(k < l && nums[l] ==  nums[l+1]) l--;
                        
                    } else if(sum > target){
                        l--;
                    } else {
                        k++;
                    }            
                }
            }
        }

        return result;
    }
}

File: 0018-4sum/0018-4sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        n = len(nums)
        result = []

        for i in range(n):
            if i != 0 and nums[i] ==  nums[i-1]:
                continue
            
            j = i+1
            for j in range(i+1,n-1):
                if j > i+1 and nums[j] == nums[j-1]:
                    continue
                
                k = j+1
                l = n-1

                while k < l:
                    sum = nums[i] + nums[j] + nums[k] + nums[l]

                    if sum == target:
                        result.append([nums[i],nums[j],nums[k],nums[l]])
                        k = k +1
                        l = l-1
                        while k < l and nums[k] == nums[k-1]:
                            k = k+1
                        while k < l and nums[l] == nums[l+1]:
                            l = l-1
                    elif sum < target:
                        k = k+1
                    else:
                        l = l-1
            

        return result
            

